src/modules/kpi-reporting/lib/mapper.js

```diff
@@ -0,0 +1,39 @@
+'use-strict';
+
+/**
+ * Maps users regisatration repo by application counting the total registrations of all time
+ * and listing the total registrations by month for the current year.
+ * @param {Array} list of registrations counted and grouped by month, application and year
+ * @return {Array} dataList list of objects that takes the form of the initialValue variable
+ */
+
+const mapRegistrations = (dataList) => {
+  const initialValue =
+    [
+      { application: 'water_vml', allTime: { registrations: 0 }, currentYear: { registrations: 0, monthly: [] } },
+      { application: 'water_admin', allTime: { registrations: 0 }, currentYear: { registrations: 0, monthly: [] } }
+    ];
+
+  return dataList.reduce((acc, row) => (
+    [{
+      application: 'water_vml',
+      allTime: { registrations: acc[0].allTime.registrations + (row.application === 'water_vml' ? row.registrations : 0) },
+      currentYear: {
+        registrations: acc[0].currentYear.registrations + (row.application === 'water_vml' && row.current_year ? row.registrations : 0),
+        monthly: (row.application === 'water_vml' && row.current_year
+          ? [...acc[0].currentYear.monthly, { month: row.month, year: row.year, registrations: row.registrations }] : [...acc[0].currentYear.monthly])
+      }
+    },
+    {
+      application: 'water_admin',
+      allTime: { registrations: acc[1].allTime.registrations + (row.application === 'water_admin' ? row.registrations : 0) },
+      currentYear: {
+        registrations: acc[1].currentYear.registrations + (row.application === 'water_admin' && row.current_year ? row.registrations : 0),
+        monthly: (row.application === 'water_admin' && row.current_year
+          ? [...acc[1].currentYear.monthly, { month: row.month, year: row.year, registrations: row.registrations }] : [...acc[1].currentYear.monthly])
+      }
+    }]
+  ), initialValue);
+};
+
+exports.mapRegistrations = mapRegistrations;
```

Comment:
I think this could be easier to understand if separated out into steps, e.g.:
```
'use-strict';
const { groupBy, mapValues } = require('lodash');

/**
 * Sums the total number of registrations in all rows
 * @param {Array} rows
 * @return {Number}
 */
const getRegistrationCount = rows => rows.reduce((acc, row) => acc + row.registrations, 0);

/**
 * Maps a group of application rows into groups for all time/current year
 * @param {Array} arr - a list of applications by month
 * @param {String} application - the application name
 * @return {Object} the data in the desired shape
 */
const mapGroup = (arr, application) => {
  const currentYear = arr.filter(row => row.current_year);
  return {
    application,
    allTime: {
      registrations: getRegistrationCount(arr)
    },
    currentYear: {
      registrations: getRegistrationCount(currentYear),
      monthly: currentYear
    }
  };
};

const filterByApplications = arr =>
  arr.filter(row => ['water_vml', 'water_admin'].includes(row.application));

/**
 * Maps users regisatration repo by application counting the total registrations of all time
 * and listing the total registrations by month for the current year.
 * @param {Array} list of registrations counted and grouped by month, application and year
 * @return {Array} dataList list of objects that takes the form of the initialValue variable
 */

const mapRegistrations = (dataList) => {
  // Filter only rows in applications water_vml or water_admin
  const filteredData = filterByApplications(dataList);

  // Group by application name
  const groups = groupBy(filteredData, row => row.application);

  // Map each group
  const mappedGroups = mapValues(groups, mapGroup);

  // Return as array
  return Object.values(mappedGroups);
};

exports.mapRegistrations = mapRegistrations;

```

Reply:
Thanks for looking at this James. I went for an approach to reduce the number of iterations over the dataset returned by the database. However I have amended the SQL to reduce the number of records returned by removing the year in the group by statement and I have also added the application names in the where clause to remove the filterByApplications line.